File: Data Structures & Algorithms/islands-and-treasure/submission-1.py
class Solution:
    def islandsAndTreasure(self, grid: List[List[int]]) -> None:
        ROWS, COLS = len(grid), len(grid[0])
        directions = [[0,1], [0,-1],[1,0],[-1,0]]
        q = deque()
        INF = 2147483647

        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == 0:
                    q.append([r, c])

        dist = 0
        while q:
            dist += 1
            for i in range(len(q)):
                r, c = q.popleft()
                for dr, dc in directions:
                    row, col = dr + r, dc + c

                    if (row < 0 or row >= ROWS or col < 0 or col >= COLS or grid[row][col] != INF):
                        continue
                    grid[row][col] = dist
                    q.append([row, col])

        return None

File: Data Structures & Algorithms/surrounded-regions/submission-1.cpp
class Solution {
public:
    void dfs(vector<vector<char>>& board, int r, int c) {
        if (r < 0 || c < 0 || r >= board.size() || c >= board[0].size() || board[r][c] != 'O') return;
        board[r][c] = '#';
        dfs(board, r + 1, c);
        dfs(board, r - 1, c);
        dfs(board, r, c + 1);
        dfs(board, r, c - 1);
    }

    void solve(vector<vector<char>>& board) {
       if (board.empty() || board[0].empty()) return;

       int ROWS = board.size(), COLS = board[0].size();

       for (int i = 0; i < ROWS; i++) {
           dfs(board, i, 0);
           dfs(board, i, COLS - 1);
       }
       for (int j = 0; j < COLS; j++) {
           dfs(board, 0, j);
           dfs(board, ROWS - 1, j);
       }

       for (int row = 0; row < ROWS; row++){
        for (int col = 0; col < COLS; col++){
            if (board[row][col] == 'O') board[row][col] = 'X';
            if (board[row][col] == '#') board[row][col] = 'O';
        }
       } 
    }
};